### ISSUE TITLE
Not able to run tests parallely

### ISSUE THREAD

**Issue by sparepal** [MarathonLabs/marathon#367]
I am new to this so please try giving me explanation in layman's terms.
I am trying to execute tests parallely. I have given the config in Marathonfile as follows:

name: "sample-app tests"
outputDir: "./marathon"
poolingStrategy:
- type: "omni"
- type: "device-model"
- type: "os-version"
- type: "manufacturer"
- type: "abi"
- shardingStrategy:
  type: "count"
  count: 2
- vendorConfiguration:
  type: "Android"
  androidSdk: "/local/android"
  applicationApk: "kotlin-buildscript/build/outputs/apk/debug/kotlin-buildscript-debug.apk"
  testApplicationApk: "kotlin-buildscript/build/outputs/apk/androidTest/debug/kotlin-buildscript-debug-androidTest.apk"
  autoGrantPermission: true

It is running the tests on the emulator but only on one of the emulators from the 3 that are launched. 
What am I missing?

**Comment by BorisOsipov** [comment]
Hi @sparepal 

Look into https://github.com/Malinskiy/marathon/issues/348 and mentioned issues. Hope it will help

**Comment by sparepal** [comment]
Hi Boris, 
Thanks for the quick reply. 

I am already on version marathon-0.5.2. 
Now, I am launching the emulators(manually) using -prop ro.serialno XYZ(uniquely for each)

Still it only runs on single emulator because the ro.serialno does set to XYZ. Do you know how to fix that?

**Comment by Malinskiy** [comment]
Hey @sparepal 
Can you verify that you actually changed the serial number of the device? You can do this either by using `adb shell getprop` or opening the settings app on the device and checking the serial number there.

It also looks a bit weird that you're using all of the pooling strategies: you'll need a lot of devices for each of those pools. You can't use "omni" pool with the other ones btw. "omni" means all of the devices are used together in a group.
When you specify `type: "os-version"` then you're asking marathon to execute your tests on all the OS versions of the connected devices separately. For example, you have one API 25 device and one API 29 device connected. Marathon will execute all your tests twice: once on API25 device and once on API29. If you add more devices with the same API version then marathon will parallelize between them.

When you specify multiple pooling strategies I believe the behaviour will be `the-last-one-wins`. So try to use "omni" for starters (it is a default one).

Also please use `` block for YAML snippets: it's hard to catch indentation errors otherwise.

**Comment by sparepal** [comment]
Thanks for the clarification on the config, will use the default omni!
About the serial number, I used adb shell getprop after launching using emulator@ <name> .prop ro.serialno=123456. It does not reflect in the [ro.serialno] when I hit the getprop command or view from the settings app. 
I'm sure not reflecting is the problem since it's not able to differentiate between the emulators launched.

**Comment by Malinskiy** [comment]
If you have root permissions to the device then you can set marathon's custom serialno property and it will be automatically picked up
```
adb -s $DEVICE shell su root setprop marathon.serialno $SERIAL
```

FYI:
It used to be possible to set the emulator's serial number before. Unfortunately, it doesn't work anymore. The way the serial is working is during the kernel startup the `androidboot.serialno` kernel argument is read and it populates the `ro.boot.serialno` and `ro.serialno` properties. Previously, the `-append` function for kernel parameters of qemu would override the default serial if you would pass the `androidboot.serialno` argument again. This isn't the case anymore. If you need this to work - please file a bug for the Android Emulator project or ask them for an alternative way to set the serial. You can ofc rebuild the whole image with a new serial but that doesn't sound as a scalable solution in my opinion.

A snippet of how it worked before in the past is below:
```
emulator -avd Pixel_2_API_25 -qemu -append 'androidboot.serialno=1234567899'
```

**Comment by sparepal** [comment]
Works, now! 
Thanks Malinskiy!

**Comment by syne10g** [comment]
@sparepal 
thank you for the question, the solution I applied and it works.
But I ran into a problem. I run 4 emulators, and only one works, the rest give an error on my app:
`app won't run without google play services which are not supported by your device`
have you ever seen anything like this?

**Comment by Malinskiy** [comment]
@syne10g  it seems like you're using a version of the android emulator that doesn't have the google play services installed, e.g. `system-images;android-29;google_apis;x86_64` instead of `system-images;android-29;default;x86_64`.
